bpo-39950: add `pathlib.Path.hardlink_to()` method that supersedes `link_to()`

barneygale opened the pull request:
The argument order of `link_to()` is reversed compared to what one may expect, so:

    a.link_to(b)

Might be expected to create *a* as a link to *b*, in fact it creates *b* as a link to *a*, making it function more like a "link from". This doesn't match `symlink_to()` nor the documentation and doesn't seem to be the original author's intent.

This PR deprecates `link_to()` and introduces `hardlink_to()`, which has the same argument order as `symlink_to()`.

<!-- issue-number: [bpo-39950](https://bugs.python.org/issue39950) -->
https://bugs.python.org/issue39950
<!-- /issue-number -->

Automerge-Triggered-By: GH:brettcannon

brettcannon commented on Doc/library/pathlib.rst:
```diff
@@ -1097,14 +1106,6 @@ call fails (for example because the path doesn't exist).
-
-   Create a hard link pointing to a path named *target*.
-
-   .. versionchanged:: 3.8
```
Don't delete the documentation as preexisting code will be referencing this method and thus people will want to know what it does and how it works. I would add a deprecation notice and point people to `hardlink_to()`.

brettcannon commented on Doc/library/pathlib.rst:
```diff
@@ -1073,6 +1073,15 @@ call fails (for example because the path doesn't exist).
+
+   Make this path a hard link to the same file as *target*.
+
+   .. note::
```
This note is unnecessary if this is the recommended way to to do things. Instead, have `link_to()` point to this method.

barneygale replied:
I'm duplicating the note from `Path.symlink_to()`, which notes the difference in argument order from `os.symlink()`. If that's worth calling out in `symlink_to()`, is it not worth also calling out in `link_to()`?

brettcannon replied:
Fair enough. I think my brain read it as referencing `Path.link_to()`.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -417,10 +417,10 @@ def lchmod(self, pathobj, mode):
     unlink = os.unlink
 
     if hasattr(os, "link"):
-        link_to = os.link
```
This is never publicly exposed, correct? Same with the rename down below?

pitrou replied:
Right, this is an implementation detail.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -1361,13 +1361,23 @@ def lstat(self):
         return self._accessor.lstat(self)
 
-    def link_to(self, target):
+    def link_to(self, link):
```
You can't change the argument name without continuing support for the old name as that's an API change.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -1361,13 +1361,23 @@ def lstat(self):
-        Create a hard link pointing to a path named target.
+        Make *link* a hard link pointing to the same file as this path.
+
+        Deprecated since Python 3.9, use `hardlink_to()`.
```
Where is the `DeprecationWarning`? Also need to specify when removal is planned (either 3.11 or 4.0 depending on how bad it will be to keep it around).

barneygale replied:
I'm not sure whether this would be removed in 3.11 or 4.0 - how can I work that out?

brettcannon replied:
@barneygale Have a discussion on the issue.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -1361,13 +1361,23 @@ def lstat(self):
+
+    def hardlink_to(self, target):
+        """
+        Make this path a hard link pointing to the same file as the given path.
```
Probably worth adding a note like in `symlink_to()`'s docstring that the order is reversed as compared to `os.link()`.

brettcannon reviewed (changes requested):
Some doc restructuring and also making sure the public API doesn't break due to this.

barneygale committed:
bpo-39925: add `pathlib.Path.hardlink_to()` method that supersedes `link_to()`

The argument order of `link_to()` is reversed, so:

    a.link_to(b)

Might be expected to create *a* as a link to *b*, in fact it creates *b*
as a link to *a*. This doesn't match `symlink_to()` nor the documentation
and doesn't seem to be the original author's intent.

This commit deprecates `link_to()` and introduces `hardlink_to()`, which
has the same argument order as `symlink_to()`.

pitrou commented on Lib/test/test_pathlib.py:
```diff
@@ -1823,6 +1824,24 @@ def test_link_to(self):
+        link.hardlink_to(target)
+        self.assertEqual(link.stat().st_size, size)
+        self.assertTrue(os.path.samefile(target, link))
+        self.assertTrue(target.stat)
```
I realize you mostly copied this test from above, but since `target.stat` is always a method, it will always be true. This seems to be a mistake in the tests.
(FTR, in the first iterations of the PEP design, `Path.stat` was a cached property, I think these checks are a remnant of this previous design)

pitrou commented on Lib/pathlib.py:
```diff
@@ -1363,11 +1364,26 @@ def lstat(self):
+
+    def hardlink_to(self, target):
+        """
+        Make this path a hard link pointing to the same file as the given path.
```
`to the same file as *target*`, perhaps?

pitrou reviewed (commented):
This looks mostly good. Just two comments.

Also, can you rebase/merge from master?

barneygale committed:
Add info to "What's New" document
